In [2]:
import pandas as pd
import numpy as np

### Задание 1.
Выберите датасет, который имеет отношение к вашей области интересов или исследований. Датасет должен содержать неструктурированные данные, требующие разметки для решения конкретной задачи, например, анализа настроений или распознавания именованных сущностей.

In [3]:
# Загрузка данных
books = pd.read_csv('C:/Users/HYPERPC/Desktop/IT/Data_collection/Homework/DZ_9/Books.csv')
ratings = pd.read_csv('C:/Users/HYPERPC/Desktop/IT/Data_collection/Homework/DZ_9/Ratings.csv')
users = pd.read_csv('C:/Users/HYPERPC/Desktop/IT/Data_collection/Homework/DZ_9/Users.csv')

C:\Users\HYPERPC\AppData\Local\Temp\ipykernel_5624\1946110222.py:2: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv('C:/Users/HYPERPC/Desktop/IT/Data_collection/Homework/DZ_9/Books.csv')


In [4]:
# Проверка типов данных
print(books.dtypes)

ISBN                   object
Book-Title             object
Book-Author            object
Year-Of-Publication    object
Publisher              object
Image-URL-S            object
Image-URL-M            object
Image-URL-L            object
dtype: object


In [5]:
# Проверка уникальных значений в проблемном столбце
print(books.iloc[:, 3].unique())

[2002 2001 1991 1999 2000 1993 1996 1988 2004 1998 1994 2003 1997 1983
 1979 1995 1982 1985 1992 1986 1978 1980 1952 1987 1990 1981 1989 1984 0
 1968 1961 1958 1974 1976 1971 1977 1975 1965 1941 1970 1962 1973 1972
 1960 1966 1920 1956 1959 1953 1951 1942 1963 1964 1969 1954 1950 1967
 2005 1957 1940 1937 1955 1946 1936 1930 2011 1925 1948 1943 1947 1945
 1923 2020 1939 1926 1938 2030 1911 1904 1949 1932 1928 1929 1927 1931
 1914 2050 1934 1910 1933 1902 1924 1921 1900 2038 2026 1944 1917 1901
 2010 1908 1906 1935 1806 2021 '2000' '1995' '1999' '2004' '2003' '1990'
 '1994' '1986' '1989' '2002' '1981' '1993' '1983' '1982' '1976' '1991'
 '1977' '1998' '1992' '1996' '0' '1997' '2001' '1974' '1968' '1987' '1984'
 '1988' '1963' '1956' '1970' '1985' '1978' '1973' '1980' '1979' '1975'
 '1969' '1961' '1965' '1939' '1958' '1950' '1953' '1966' '1971' '1959'
 '1972' '1955' '1957' '1945' '1960' '1967' '1932' '1924' '1964' '2012'
 '1911' '1927' '1948' '1962' '2006' '1952' '1940' '1951' '1931' '1954

In [6]:
# Функция для преобразования значений в целые числа
def to_integer(value):
    try:
        return int(value)
    except ValueError:
        return np.nan

# Применение функции к столбцу и преобразование типа данных
books['Year-Of-Publication'] = books['Year-Of-Publication'].apply(to_integer).astype('Int64')

# Проверка преобразования типов данных
print(books.dtypes)

ISBN                   object
Book-Title             object
Book-Author            object
Year-Of-Publication     Int64
Publisher              object
Image-URL-S            object
Image-URL-M            object
Image-URL-L            object
dtype: object


In [7]:
# Первичный анализ
print(books.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 8 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271358 non-null  object
 3   Year-Of-Publication  271357 non-null  Int64 
 4   Publisher            271358 non-null  object
 5   Image-URL-S          271360 non-null  object
 6   Image-URL-M          271360 non-null  object
 7   Image-URL-L          271357 non-null  object
dtypes: Int64(1), object(7)
memory usage: 16.8+ MB
None


In [8]:
print(books.head())

         ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author  Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                 2002     Oxford University Press   
1  Richard Bruce Wright                 2001       HarperFlamingo Canada   
2          Carlo D'Este                 1991             HarperPerennial   
3      Gina Bari Kolata                 1999        Farrar Straus Giroux   
4       E. J. W. Barber                 1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazo

In [9]:
print(ratings.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149780 entries, 0 to 1149779
Data columns (total 3 columns):
 #   Column       Non-Null Count    Dtype 
---  ------       --------------    ----- 
 0   User-ID      1149780 non-null  int64 
 1   ISBN         1149780 non-null  object
 2   Book-Rating  1149780 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 26.3+ MB
None


In [10]:
print(ratings.head())

   User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6


In [11]:
print(users.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   User-ID   278858 non-null  int64  
 1   Location  278858 non-null  object 
 2   Age       168096 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB
None


In [12]:
print(users.head())

   User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN


Очистка данных

Проверка и обработка пропущенных значений:

In [13]:
# Проверка пропущенных значений
print(books.isnull().sum())

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    3
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64


In [14]:
print(ratings.isnull().sum())

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64


In [15]:
print(users.isnull().sum())

User-ID          0
Location         0
Age         110762
dtype: int64


In [16]:
#В данном случае, столбцы с URL изображений можно удалить, так как они не имеют значения для анализа.
books = books.drop(columns=['Image-URL-S', 'Image-URL-M', 'Image-URL-L'])

In [17]:
# Заполним пропуски значением "Unknown"
books['Book-Author'] = books['Book-Author'].fillna('Unknown')
books['Publisher'] = books['Publisher'].fillna('Unknown')

In [18]:
# Найдем моду (наиболее часто встречающееся значение) для столбца Year-Of-Publication
mode_year = books['Year-Of-Publication'].mode()[0]

# Заполним пропуски модой
books['Year-Of-Publication'] = books['Year-Of-Publication'].fillna(mode_year)

In [19]:
print(books.isnull().sum())

ISBN                   0
Book-Title             0
Book-Author            0
Year-Of-Publication    0
Publisher              0
dtype: int64


In [20]:
# Описательная статистика
users.describe()

User-ID            Age
count  278858.00000  168096.000000
mean   139429.50000      34.751434
std     80499.51502      14.428097
min         1.00000       0.000000
25%     69715.25000      24.000000
50%    139429.50000      32.000000
75%    209143.75000      44.000000
max    278858.00000     244.000000

In [21]:
# Найдем значения возраста больше 100
ages_above_100 = users[users['Age'] > 100]['Age']

# Выведем эти значения
print(ages_above_100)

1288      103.0
1322      104.0
1578      231.0
3084      104.0
3210      119.0
          ...  
276047    127.0
276352    104.0
277107    104.0
277503    103.0
278301    104.0
Name: Age, Length: 366, dtype: float64


Мы заменим значения возраста, превышающие 100 лет, а также пропущенные значения (NaN), на медиану возраста, равную 32 годам. Поскольку возраст свыше 100 лет является аномальными данными (выбросами), использование медианы позволит улучшить качество данных и сделать их более валидными для дальнейшего анализа.

In [22]:
# Найдем медиану возраста
age_median = users['Age'].median()

# Заменим выбросы (возраст > 100) и пропущенные значения на медиану
users['Age'] = users['Age'].apply(lambda x: age_median if x > 100 or pd.isnull(x) else x)

# Проверим результат
print(users['Age'].describe())

count    278858.000000
mean         33.544844
std          10.789130
min           0.000000
25%          29.000000
50%          32.000000
75%          35.000000
max         100.000000
Name: Age, dtype: float64


In [23]:
print(users.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   User-ID   278858 non-null  int64  
 1   Location  278858 non-null  object 
 2   Age       278858 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 6.4+ MB
None


In [24]:
users['Age'] = users['Age'].round(0).astype(int)

print(users.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 columns):
 #   Column    Non-Null Count   Dtype 
---  ------    --------------   ----- 
 0   User-ID   278858 non-null  int64 
 1   Location  278858 non-null  object
 2   Age       278858 non-null  int32 
dtypes: int32(1), int64(1), object(1)
memory usage: 5.3+ MB
None


In [25]:
# Удаление строк с пропусками:
books = books.dropna()
ratings = ratings.dropna()
users = users.dropna()

In [26]:
# Проверка дубликатов
print(books.duplicated().sum())
print(ratings.duplicated().sum())
print(users.duplicated().sum())

# Удаление дубликатов
books = books.drop_duplicates()
ratings = ratings.drop_duplicates()
users = users.drop_duplicates()

0
0
0


Мы объединим данные из различных файлов, используя общие столбцы, такие как bookid и userid. Это позволит нам создать единую таблицу, содержащую все необходимые данные для дальнейшего анализа и обработки

In [27]:
print(books.info())
print(ratings.info())
print(users.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 271360 entries, 0 to 271359
Data columns (total 5 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ISBN                 271360 non-null  object
 1   Book-Title           271360 non-null  object
 2   Book-Author          271360 non-null  object
 3   Year-Of-Publication  271360 non-null  Int64 
 4   Publisher            271360 non-null  object
dtypes: Int64(1), object(4)
memory usage: 10.6+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1149780 entries, 0 to 1149779
Data columns (total 3 columns):
 #   Column       Non-Null Count    Dtype 
---  ------       --------------    ----- 
 0   User-ID      1149780 non-null  int64 
 1   ISBN         1149780 non-null  object
 2   Book-Rating  1149780 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 26.3+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 278858 entries, 0 to 278857
Data columns (total 3 c

In [28]:
# Использование внутреннего соединения (inner join)
final_data = pd.merge(pd.merge(books, ratings, on='ISBN'), users, on='User-ID', how='inner')

# Проверим результат
print(final_data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1031136 entries, 0 to 1031135
Data columns (total 9 columns):
 #   Column               Non-Null Count    Dtype 
---  ------               --------------    ----- 
 0   ISBN                 1031136 non-null  object
 1   Book-Title           1031136 non-null  object
 2   Book-Author          1031136 non-null  object
 3   Year-Of-Publication  1031136 non-null  Int64 
 4   Publisher            1031136 non-null  object
 5   User-ID              1031136 non-null  int64 
 6   Book-Rating          1031136 non-null  int64 
 7   Location             1031136 non-null  object
 8   Age                  1031136 non-null  int32 
dtypes: Int64(1), int32(1), int64(2), object(5)
memory usage: 67.9+ MB
None


In [29]:
final_data.to_csv('final_data_labeled.csv', index=False)

### Задание 2.
Выполните разметку на основе правил (rule-based labeling) на подмножестве выбранного датасета. Разработайте и реализуйте набор правил или условий, которые позволят автоматически присваивать метки данным на основе определенных шаблонов или критериев.

In [30]:
# Создание фиктивного DataFrame
np.random.seed(42)  # Для воспроизводимости результатов

n = 100  # Количество строк в DataFrame

# Генерация случайных данных для DataFrame
data = {
    'Year-Of-Publication': np.random.randint(1950, 2023, size=n),  # Год публикации книги
    'Book-Rating': np.random.randint(1, 11, size=n),  # Рейтинг книги
    'Age': np.random.randint(10, 80, size=n),  # Возраст пользователя
    'Page-Count': np.random.randint(50, 1000, size=n),  # Количество страниц в книге
    'Genre': np.random.choice(['Science Fiction', 'Fantasy', 'Mystery', 'Thriller', 'Romance', 'Drama', 'Non-Fiction'], size=n),  # Жанр книги
    'Review-Count': np.random.randint(0, 100, size=n)  # Количество отзывов на книгу
}

# Создание DataFrame из сгенерированных данных
final_data_fictitious = pd.DataFrame(data)

# Функции для создания меток на основе правил

def label_row(row):
    return pd.Series({
        'Year-Label': 'Before 2000' if row['Year-Of-Publication'] < 2000 else 'After 2000',
        'Rating-Label': 'High' if row['Book-Rating'] >= 7 else 'Medium' if row['Book-Rating'] >= 4 else 'Low',
        'Age-Label': 'Child' if row['Age'] < 18 else 'Adult' if row['Age'] < 65 else 'Senior',
        'Page-Count-Label': 'Short' if row['Page-Count'] < 100 else 'Medium' if row['Page-Count'] < 300 else 'Long',
        'Genre-Label': 'Speculative Fiction' if row['Genre'] in ['Science Fiction', 'Fantasy']
        else 'Suspense' if row['Genre'] in ['Mystery', 'Thriller']
        else 'Emotional' if row['Genre'] in ['Romance', 'Drama']
        else 'Other',
        'Review-Count-Label': 'Few Reviews' if row['Review-Count'] < 10 else 'Moderate Reviews' if row['Review-Count'] < 50 else 'Many Reviews'
    })

# Применение функции для создания меток ко всем строкам DataFrame
labels = final_data_fictitious.apply(label_row, axis=1)

# Добавление новых меток в DataFrame
final_data_fictitious = pd.concat([final_data_fictitious, labels], axis=1)

# Вывод первых нескольких строк DataFrame для проверки результатов
print(final_data_fictitious[['Year-Of-Publication', 'Year-Label', 
                             'Book-Rating', 'Rating-Label', 
                             'Age', 'Age-Label',
                             'Page-Count', 'Page-Count-Label',
                             'Genre', 'Genre-Label',
                             'Review-Count', 'Review-Count-Label']].head())

# Сохранение DataFrame в CSV-файл
final_data_fictitious.to_csv('final_data_fictitious.csv', index=False)

   Year-Of-Publication   Year-Label  Book-Rating Rating-Label  Age Age-Label  \
0                 2001   After 2000            8         High   12     Child   
1                 1964  Before 2000            5       Medium   29     Adult   
2                 2021   After 2000            3          Low   68    Senior   
3                 2010   After 2000            8         High   45     Adult   
4                 1970  Before 2000            6       Medium   28     Adult   

   Page-Count Page-Count-Label            Genre          Genre-Label  \
0         921             Long          Romance            Emotional   
1         735             Long            Drama            Emotional   
2         841             Long      Non-Fiction                Other   
3         675             Long      Non-Fiction                Other   
4         337             Long  Science Fiction  Speculative Fiction   

   Review-Count Review-Count-Label  
0            34   Moderate Reviews  
1           

In [31]:
final_data_fictitious.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Year-Of-Publication  100 non-null    int32 
 1   Book-Rating          100 non-null    int32 
 2   Age                  100 non-null    int32 
 3   Page-Count           100 non-null    int32 
 4   Genre                100 non-null    object
 5   Review-Count         100 non-null    int32 
 6   Year-Label           100 non-null    object
 7   Rating-Label         100 non-null    object
 8   Age-Label            100 non-null    object
 9   Page-Count-Label     100 non-null    object
 10  Genre-Label          100 non-null    object
 11  Review-Count-Label   100 non-null    object
dtypes: int32(5), object(7)
memory usage: 7.5+ KB


In [32]:
# Определение соотношения и размера подмножества для ручной разметки
split_ratio = 0.2  # 20% данных для ручной разметки

# Выбор подмножества данных с помощью параметра frac
manual_subset_df = final_data_fictitious.sample(frac=split_ratio, random_state=42)

# Сохранение подмножества данных в формате CSV для загрузки в Label Studio
manual_subset_df.to_csv('manual_subset.csv', index=False)

In [33]:
# Экспортированный файл - 'annotation_result.csv'
manual_annotations = pd.read_csv('C:/Users/HYPERPC/Desktop/IT/Data_collection/Homework/DZ_9/annotation_result.csv')

### Задача 4.
Объедините данные, размеченные вручную, с данными, размеченными на основе правил. Объедините два подмножества размеченных данных в один набор данных, сохранив при этом соответствующую структуру и целостность.

In [34]:
# Подгрузка двух наборов данных
final_data_fictitious = pd.read_csv('final_data_fictitious.csv')
manual_annotations = pd.read_csv('annotation_result.csv')

# Объединение данных
combined_data = pd.concat([final_data_fictitious, manual_annotations], ignore_index=True)

# Сохранение объединенного набора данных (опционально)
combined_data.to_csv('combined_data.csv', index=False)

### Задача 5.
Обучите модель машинного обучения, используя объединенный набор размеченных данных. Разделите датасет на обучающий и тестовый наборы и используйте обучающий набор для обучения модели.

In [36]:
from sklearn.model_selection import train_test_split  # используется для разделения данных на обучающую и тестовую выборки
from sklearn.preprocessing import LabelEncoder, StandardScaler  # используется (LabelEncoder) для преобразования категориальных признаков в числовые, (StandardScaler) для нормализации данных.
from sklearn.ensemble import RandomForestClassifier  # используется для создания модели случайного леса.
from sklearn.linear_model import LogisticRegression  # используется для создания модели логистической регрессии.
from sklearn.metrics import accuracy_score  # используется для оценки точности моделей.

In [37]:
# Загрузка данных из CSV файла
combined_data = pd.read_csv('combined_data.csv')

# Заполнение отсутствующих данных
# Используем метод forward-fill (ffill) и backward-fill (bfill) для заполнения пропущенных значений.
combined_data.fillna(method='ffill', inplace=True)
combined_data.fillna(method='bfill', inplace=True)

# Преобразование категориальных данных в числовые с помощью LabelEncoder
# LabelEncoder преобразует категориальные текстовые данные в числовые значения.
label_encoders = {}
categorical_columns = combined_data.select_dtypes(include=['object']).columns

for column in categorical_columns:
    le = LabelEncoder()
    combined_data[column] = le.fit_transform(combined_data[column])
    label_encoders[column] = le

# Проверка наличия 'Book-Rating' перед разделением данных
if 'Book-Rating' not in combined_data.columns:
    raise ValueError("'Book-Rating' столбец отсутствует в данных")

# Разделение датасета на признаки (X) и целевую переменную (y)
X = combined_data.drop('Book-Rating', axis=1)
y = combined_data['Book-Rating']

# Разделение данных на обучающую и тестовую выборки
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Нормализация признаков с помощью StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# Вывод некоторых данных для проверки
print("Пример X_train после нормализации:\n", X_train[:5])
print("Пример y_train:\n", y_train[:5])

Пример X_train после нормализации:
 [[ 0.71363883  1.67085802 -0.25046926  1.32377676  0.30204712 -1.32017315
  -1.17815946  1.8973666  -0.46209221  1.32374538 -0.5805578  -0.24413654
  -0.38576235  0.         -0.38576235 -0.31955407 -0.38576235  0.33844201
  -0.16903085 -0.32163376 -0.11651035 -0.38576235  0.23440362]
 [-1.49494254 -1.6442943   0.8723959  -1.12011879  0.16575324  0.7574764
  -1.17815946  0.63245553 -0.46209221  0.49316004 -0.5805578  -0.24413654
  -0.38576235  0.         -0.38576235 -0.31955407 -0.38576235  0.33844201
  -0.16903085 -0.32163376 -0.11651035 -0.38576235  0.23440362]
 [-0.67213771  0.85482053 -0.52939755  0.83499765 -0.58386311  0.7574764
  -1.17815946 -0.63245553 -0.46209221  0.49316004  1.10833761 -0.24413654
  -0.38576235  0.         -0.38576235 -0.31955407 -0.38576235  0.33844201
  -0.16903085 -0.32163376 -0.11651035 -0.38576235  0.23440362]
 [-0.36899909  1.36484396  0.42181944  0.83499765  0.16575324  0.7574764
   1.25416974  1.8973666  -0.46209221 

C:\Users\HYPERPC\AppData\Local\Temp\ipykernel_5624\2842912488.py:6: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  combined_data.fillna(method='ffill', inplace=True)
C:\Users\HYPERPC\AppData\Local\Temp\ipykernel_5624\2842912488.py:7: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  combined_data.fillna(method='bfill', inplace=True)


In [38]:
# Обучение модели RandomForestClassifier
# RandomForestClassifier - это ансамблевый метод, который строит множество деревьев решений и объединяет их результаты.
rf_model = RandomForestClassifier(random_state=42)
rf_model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [39]:
# Обучение модели LogisticRegression
# LogisticRegression - это метод классификации, который моделирует вероятность принадлежности к одному из классов.
logistic_model = LogisticRegression(random_state=42, max_iter=1000)
logistic_model.fit(X_train, y_train)

LogisticRegression(max_iter=1000, random_state=42)

### Задача 6.
Оценить эффективность обученной модели на тестовом датасете. Используйте подходящие метрики оценки. Интерпретируйте результаты и проанализируйте эффективность модели в решении задачи разметки.

In [40]:
# Предсказание значений на тестовой выборке с использованием RandomForestClassifier
# Используем обученную модель для предсказания классов на тестовых данных.
y_pred_rf = rf_model.predict(X_test)

# Оценка точности модели RandomForestClassifier с использованием метрики accuracy_score
# accuracy_score вычисляет долю правильных предсказаний.
accuracy_rf = accuracy_score(y_test, y_pred_rf)
print(f'Random Forest Accuracy: {accuracy_rf}')


# Предсказание значений на тестовой выборке с использованием LogisticRegression
# Используем обученную модель для предсказания классов на тестовых данных.
y_pred_logistic = logistic_model.predict(X_test)

# Оценка точности модели LogisticRegression с использованием метрики accuracy_score
accuracy_logistic = accuracy_score(y_test, y_pred_logistic)
print(f'Logistic Regression Accuracy: {accuracy_logistic}')

Random Forest Accuracy: 0.5416666666666666
Logistic Regression Accuracy: 0.4166666666666667


Метод Random Forest показал более высокую точность (54.17%) по сравнению с Logistic Regressio (41.67%). 
Следовательно, модель Random Forest лучше справляется с поставленной задачей на данном наборе данных.

*Сравнение моделей:* 
На основании этих результатов можно сделать вывод, что модель Random Forest демонстрирует лучшую производительность по сравнению с моделью Logistic Regression на данном наборе данных. Разница в точности составляет 12.5%, что может свидетельствовать о более эффективной работе модели Random Forest в текущих условиях задачи.

*Абсолютные значения точности:*
В то же время, следует отметить, что обе модели показывают достаточно низкую точность. Это может указывать на несколько возможных причин, такие как:
1. Сложность данных для предсказания.
2. Необходимость дополнительной настройки.
3. Качество данных и признаки.

Дальнейшие действия:
1. Анализ данных: Необходимо провести более детальный анализ данных, чтобы выявить возможные проблемы и аномалии.
2. Улучшение признаков: Следует проверить возможности для создания новых признаков, а также устранения ненужных или малоинформативных признаков.
3. Пробовать другие модели: Стоит рассмотреть использование других моделей машинного обучения.
4. Кросс-валидация: Рекомендуется использовать методы кросс-валидации для более точной оценки производительности моделей и выбора наилучшей из них.